### PR TITLE
Add function to export user report data from Gophish

### PR DESCRIPTION
**Pull request by JCantu248**
#  #

## 🗣 Description ##

Function was added to gather user report data for all campaigns associated with a given assessment. The data is obtained by using the GoPhish API, and written to a JSON file for later use by PCA operators. Each campaign is written to its own JSON file. 

## 💭 Motivation and context ##

As part of the PCA team's effort to develop Live PCA, it is requested that user report data (that is targeted user clicking on links) be obtained from GoPhish and collected into a format that allows for import into the PCA database. This will allow for archival storage of assessment and campaign data. 

## 🧪 Testing ##

Code base was tested in the Staging COOL environment. Two new campaigns were set up in GoPhish running in Staging COOL, and after their completion, the GoPhish Export command was run with the appropriate assessment name.  A JSON for each campaign was generated, with properly formatted JSON. 

Unfortunately since we don't have corresponding assessment and campaign documents from GoPhish loaded into the PCA DB, I wasnt able to test importing the generating JSON. What I did instead was wrote the first_report datetime into a known good and valid user report JSON file, and forced an overwrite of the document. I did this to make sure that the datetime format I was using for first_report was matching what the importer and PCA DB were expecting. 

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
 | interface_type | Type of interface GoPhish will use with mail server. | string | "SMTP" | no |
 | ignore_cert | Indicate if GoPhish should ignore certs with mail server. | boolean | `True` | no |
 
+## User Report Field Dictionary ##
+
+User report JSONs are also exported by ```gophish-export```. User reports
```
Comment (dav3r) [inline README.md]: Single backticks are sufficient here:
```suggestion
User report JSONs are also exported by `gophish-export`. User reports
```

File: README.md [hunk 2]
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
...
+summarize data from targeted user clicks on phishing emails generated by a
+campaign. These JSON files are later imported into the PCA database.
+
+| Name | Description | Type | Default | Required |
+|------|-------------|:----:|:-------:|:--------:|
+| customer | A corresponding customer ID that this user report document is associated with. It is left null as it will be filled in later by a PCA operator or another tool prior to importation | string | null | No |
+| assessment | A corresponding assessment ID that this user report is associated with. | string | | yes |
+| campaign | A corresponding campaign ID that this user report is associated with. | string| | yes |
+| first_report | The first report (click) generated by a targeted user. Formatted by YYYY-MM-DDT: hh:mm.ss"| datetime | | yes |
+| total_num_reports | The total number of user reports received during a campaign.| integer | | no
```
Comment (dav3r) [inline README.md]: I recommend sorting these rows alphabetically by field name.

Comment (dav3r) [inline README.md]: Clean up wording a bit and remove PCA-specific language:
```suggestion
| Name | Description | Type | Default | Required |
|------|-------------|:----:|:-------:|:--------:|
| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as `null`. | string | `null` | no |
| assessment | Assessment ID that this user report is associated with. | string | | yes |
| campaign | Campaign ID that this user report is associated with. | string | | yes |
| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDT: hh:mm.ss" | datetime | | yes |
| total_num_reports | Total number of user reports received during a campaign. | integer | | no |
```

File: README.md [hunk 3]
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
...
+User report JSONs are also exported by ```gophish-export```. User reports
+summarize data from targeted user clicks on phishing emails generated by a
+campaign. These JSON files are later imported into the PCA database.
+
+| Name | Description | Type | Default | Required |
+|------|-------------|:----:|:-------:|:--------:|
+| customer | A corresponding customer ID that this user report document is associated with. It is left null as it will be filled in later by a PCA operator or another tool prior to importation | string | null | No |
+| assessment | A corresponding assessment ID that this user report is associated with. | string | | yes |
+| campaign | A corresponding campaign ID that this user report is associated with. | string| | yes |
+| first_report | The first report (click) generated by a targeted user. Formatted by YYYY-MM-DDT: hh:mm.ss"| datetime | | yes |
```
Comment (dav3r) [inline README.md]: Is `YYYY-MM-DDT: hh:mm.ss` really the correct format or should it be `YYYY-MM-DDThh:mm.ss`?

File: src/_version.py [hunk 4]
```diff
@@ -1,2 +1,2 @@
 """This file defines the version of this project."""
-__version__ = "0.0.4"
+__version__ = "0.2.0"
```
Comment (dav3r) [inline src/_version.py]: What happened to `0.1.0`?

File: README.md (hunk 2 above)
Comment (dav3r) [inline README.md]: I thought over in [this discussion](https://github.com/cisagov/pca-core/pull/5#discussion_r651079451), we were saying that `total_num_reports` cannot be blank, i.e. it is required?

File: src/tools/gophish_export.py [hunk 5]
```diff
@@ -253,6 +253,44 @@ def get_application(rawEvent):
...
+        # iterate over clicks and find the earliest click
+        for click in campaign["clicks"]:
+            if (
+                first_report is None
+                or datetime.strptime(click["time"].split(".")[0], "%Y-%m-%dT%H:%M:%S")
+                < first_report
+            ):
+                first_report = datetime.strptime(
+                    click["time"].split(".")[0], "%Y-%m-%dT%H:%M:%S"
+                )
```
Comment (dav3r) [inline src/tools/gophish_export.py]: Here's a slightly cleaner way to write this:
```suggestion
            click_time = datetime.strptime(
                click["time"].split(".")[0], "%Y-%m-%dT%H:%M:%S"
            )
            if first_report is None or click_time < first_report:
                first_report = click_time
```

File: src/tools/gophish_export.py [hunk 6]
```diff
@@ -253,6 +253,44 @@ def get_application(rawEvent):
...
+
+        user_report_doc["customer"] = None
+        user_report_doc["assessment"] = assessment_id
+        user_report_doc["campaign"] = campaign["id"]
```
Comment (dav3r) [inline src/tools/gophish_export.py]: Is there any difference between `campaign_id` and `campaign["id"]` here?  If not, I'd simply go with:
```suggestion
        user_report_doc["campaign"] = campaign_id
```

File: src/tools/gophish_export.py [hunk 7]
```diff
@@ -253,6 +253,44 @@ def get_application(rawEvent):
...
+        user_report_doc["first_report"] = datetime.strftime(
+            first_report, "%Y-%m-%dT%H:%M:%S"
+        )
+        user_report_doc["total_num_reports"] = api.campaigns.summary(
+            campaign_id=campaign_id
+        ).stats.clicked
+
+        logging.info(
+            f"Writing out user reports for all campaigns in assessment: {assessment_id}"
```
Comment (dav3r) [inline src/tools/gophish_export.py]: Since this logging message is inside the `for campaign_id in campaign_ids` loop, it should indicate which campaign is being output:
```suggestion
            f"Writing out user report for campaign {campaign_id} in assessment {assessment_id}"
```

**Review by dav3r — CHANGES_REQUESTED** [review]
I have some questions, comments, and feedback- please let me know if you have any questions about what I wrote.

File: README.md (hunk 2 above)
Comment (JCantu248) [reply README.md]: I put them in this order because this is the order I have them on in the JSON file. Are you sure you wanna change the field order?

Comment (jsf9k) [reply README.md]: The order of the fields in the JSON file shouldn't be relevant, right?  It isn't a CSV.  I think @dav3r is right here.

If part of your code depends on the order of JSON dictionary keys that probably indicates a problem, since JSON does not make any promises about such ordering.  If you parse a JSON and then re-export it the order of the keys can change depending on the hash map implementation being used internally.

Comment (JCantu248) [reply README.md]: See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

File: README.md (hunk 3 above)
Comment (JCantu248) [reply README.md]: See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

File: README.md (hunk 2 above)
Comment (JCantu248) [reply README.md]: See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

**Comment by BenBreaksThings** [comment]
Chiming in out of turn here, but many of the items above are characterized as recommendations or are instances of refactoring. Not that the code is broken, just preferential construct. Can we adopt a more ops focused mindset here and accept that the working code that isn't doing anything dangerous need not be prevented from progressing due to JSON alphabetizing?

File: src/_version.py (hunk 4 above)
Comment (JCantu248) [reply src/_version.py]: Not sure how I messed up the bump_version. I manually fixed it here: 3cd389cce6498bf6579e9ce3c47fdb0968e9febf

File: src/tools/gophish_export.py (hunk 6 above)
Comment (JCantu248) [reply src/tools/gophish_export.py]: I tried that before during my testing. The campaign_id that iterates over campaign_ids, which is in turn returned from get_campaign_ids() is an integer value. campaign["id"] is a string value. The schema for the campaign field in the user_report is defined as a string as well, as thats what campaign _ids were defined as throughout the PCA database and our legacy data. Im using the campaign["id"] so I can load string data into the current schema.

File: README.md (hunk 2 above)
Comment (dav3r) [reply README.md]: @JCantu248, https://github.com/cisagov/gophish-tools/pull/33/commits/bac2cb88e00b183bdbae8f990c3ad5a48b01ef05 does not correct the alphabetical order.

Also, please use more descriptive messages in your commits: `Merging remote` is not particularly helpful here.  We strongly encourage everyone to follow the commit message style and guidelines described in [this post](https://chris.beams.io/posts/git-commit/#seven-rules), as referenced in our [Development Guide](https://github.com/cisagov/development-guide#git-github-and-you).

File: README.md (hunk 3 above)
Comment (dav3r) [reply README.md]: It looks like you corrected the format in https://github.com/cisagov/gophish-tools/pull/33/commits/ccdf910c06d009dca6304f8223285cb2d9d99042, but then overwrote it with https://github.com/cisagov/gophish-tools/commit/bac2cb88e00b183bdbae8f990c3ad5a48b01ef05.  I assume that you did not mean to do that.  Please let me know when it's fixed.

File: README.md (hunk 2 above)
Comment (dav3r) [reply README.md]: It looks like you changed this field to be required in ccdf910, but then overwrote it with bac2cb8. I assume that you did not mean to do that. Please let me know when it's fixed.

**Comment by felddy** [comment]
> Chiming in out of turn here, but many of the items above are characterized as recommendations or are instances of refactoring. Not that the code is broken, just preferential construct. Can we adopt a more ops focused mindset here and accept that the working code that isn't doing anything dangerous need not be prevented from progressing due to JSON alphabetizing?

It's all good.  All chiming is welcome, at all times.

We do indeed have preferential constructs, and they're usually not just whims.   We strive to give our codebase as much longevity as possible.  This means that we hope that it will continue to be in useful long after we're done writing it.  The more readable, and uniform we can make our projects, the better the chance they will survive changes in maintainers and owners.

It's ok to disagree with a suggestion but we'd ask that you provide your reasoning when you do.  Our coding best-practices are always evolving and we're always looking to improve them.  

I don't think any of the conversations and suggestions above are out of the ordinary.  This is actually less of a barrage than I usually get on my PRs.   😅 

In any case, thank you for contributing and running the gauntlet with us.  It's not always fun, but the project will be better for it.

File: src/tools/gophish_export.py (hunk 6 above)
Comment (dav3r) [reply src/tools/gophish_export.py]: Thanks for the explanation.  In the interest of more readable code, how about something like this?
```
# get_campaign_ids() returns integers, but user_report_doc["campaign"]
# expects a string
user_report_doc["campaign"] = str(campaign_id)
```

File: src/tools/gophish_export.py (hunk 5 above)
Comment (JCantu248) [reply src/tools/gophish_export.py]: See 724b35ac39da24c61713cc0dd0addb7a5f3befe5

File: src/tools/gophish_export.py (hunk 6 above)
Comment (JCantu248) [reply src/tools/gophish_export.py]: See 724b35ac39da24c61713cc0dd0addb7a5f3befe5

File: README.md (hunk 2 above)
Comment (JCantu248) [reply README.md]: See cc24b7780b4ff24ca8f43633d2dce6a214d86892

Comment (dav3r) [reply README.md]: https://github.com/cisagov/gophish-tools/commit/cc24b7780b4ff24ca8f43633d2dce6a214d86892 fixes the alphabetization issue, but `total_num_reports` is still listed as optional.  Did you mean to put this comment in the conversation above?

Comment (JCantu248) [reply README.md]: Nah, it was an oversight. See 51786dc6d5b6d82de55ecbf822644fe2fe43ce71

File: README.md (hunk 3 above)
Comment (JCantu248) [reply README.md]: Fixed here 71d9fc125456322326ea57a91ce7d86fbf07d45e

Comment (dav3r) [reply README.md]: Thanks!

File: README.md [hunk 8]
```diff
@@ -21,7 +21,8 @@ campaign assessors to simplify the process of managing GoPhish campaigns.
 * `gophish-complete` - Completes a campaign in GoPhish and/or outputs a
   GoPhish campaign summary.
 * `gophish-export` - Exports all the data from an assessment within GoPhish
-  into a single JSON file.
+  into a single JSON file. In addition, user report JSONs for each campaign in
+  an assessment will also be generated, for later import into the PCA database.
```
Comment (dav3r) [inline README.md]: Remove reference to PCA- this repo should stand on it's own as a Gophish helper utility, regardless of whether or not it is used for PCAs.
```suggestion
  an assessment will also be exported.
```

File: README.md [hunk 9]
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
 | interface_type | Type of interface GoPhish will use with mail server. | string | "SMTP" | no |
 | ignore_cert | Indicate if GoPhish should ignore certs with mail server. | boolean | `True` | no |
 
+## User Report Field Dictionary ##
+
+User report JSONs are also exported by `gophish-export`. User reports
+summarize data from targeted user clicks on phishing emails generated by a
+campaign. These JSON files are later imported into the PCA database.
```
Comment (dav3r) [inline README.md]: Remove reference to PCA- this repo should stand on it's own as a Gophish helper utility, regardless of whether or not it is used for PCAs.
```suggestion
campaign.
```

File: README.md (hunk 2 above)
Comment (dav3r) [reply README.md]: Thanks @JCantu248 for fixing the alphabetization in https://github.com/cisagov/gophish-tools/pull/33/commits/cc24b7780b4ff24ca8f43633d2dce6a214d86892. 👍

File: README.md (hunk 8 above)
Comment (JCantu248) [reply README.md]: See 2d96175613954e3c341aa38614d80369113217af

File: README.md (hunk 9 above)
Comment (JCantu248) [reply README.md]: See 2d96175613954e3c341aa38614d80369113217af

File: README.md (hunk 8 above)
Comment (dav3r) [reply README.md]: Thanks!  FYI, you can use [GitHub's "Commit suggestion" feature](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) to save time when making changes like this in the future.

**Review by dav3r — APPROVED** [review]
Thanks for working through all of my feedback - this looks great now! 🎸

**Comment by dav3r** [comment]
@JCantu248 - I just noticed that you still need to pull in the latest changes from `develop`.  Please take care of that when you can.

File: src/tools/gophish_export.py [hunk 10]
```diff
@@ -253,6 +253,43 @@ def get_application(rawEvent):
...
+        user_report_doc["first_report"] = datetime.strftime(
+            first_report, "%Y-%m-%dT%H:%M:%S"
+        )
+        user_report_doc["total_num_reports"] = api.campaigns.summary(
+            campaign_id=campaign_id
+        ).stats.clicked
+
+        logging.info(
+            f"Writing out user report for campaign {campaign_id} in assessment {assessment_id}"
+        )
```
Comment (mcdonnnj) [inline src/tools/gophish_export.py:289]: Logging calls should use the `printf` style notation, but I know there are a lot of them in this repository. Please make an issue in the repo to switch all `logging` library calls to remove f-strings and use `printf` style formatting.

File: src/tools/gophish_export.py [hunk 11]
```diff
@@ -253,6 +253,43 @@ def get_application(rawEvent):
     return application
 
 
+def export_user_reports(api, assessment_id):
+    """Build and export a user_report JSON object for each campaign in an assessment."""
```
Comment (mcdonnnj) [inline src/tools/gophish_export.py]:
```suggestion
    """Build and export a user_report JSON file for each campaign in an assessment."""
```

File: src/tools/gophish_export.py [hunk 12]
```diff
@@ -253,6 +253,43 @@ def get_application(rawEvent):
...
+
+        # iterate over clicks and find the earliest click
+        for click in campaign["clicks"]:
+            click_time = datetime.strptime(
+                click["time"].split(".")[0], "%Y-%m-%dT%H:%M:%S"
+            )
+            if first_report is None or click_time < first_report:
+                first_report = click_time
+
+        user_report_doc["customer"] = None
```
Comment (mcdonnnj) [inline src/tools/gophish_export.py]: Is a `null` the desired output in the JSON for this field? Your README change indicates the field is a `string`. A `null` and a `string` are not the same, so I am suggesting we instead output an empty string for this field.

I have also added a comment to provide some insight on why such a field is in the output in the first place. @dav3r I would appreciate your input on making this comment better.

```suggestion
        # The "customer" field is a placeholder added for operator convenience when
        # working with the JSON file created.
        user_report_doc["customer"] = ""
```

File: README.md [hunk 13]
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
...
+summarize data from targeted user clicks on phishing emails generated by a
+campaign.
+
+| Name | Description | Type | Default | Required |
+|------|-------------|:----:|:-------:|:--------:|
+| assessment | Assessment ID that this user report is associated with. | string | | yes |
+| campaign | Campaign ID that this user report is associated with. | string | | yes |
+| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as `null`. | string | `null` | no |
+| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDThh:mm.ss" | datetime | | yes |
+| total_num_reports | Total number of user reports received during a campaign. | integer | | yes |
```
Comment (mcdonnnj) [inline README.md]: `Default`/`Required` doesn't make much sense for a JSON output. I also updated one of the descriptions to reflect a change I suggested in the code.

```suggestion
| Name | Description | Type |
|------|-------------|:----:|
| assessment | Assessment ID that this user report is associated with. | string |
| campaign | Campaign ID that this user report is associated with. | string |
| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as an empty string. | string |
| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDThh:mm.ss" | datetime |
| total_num_reports | Total number of user reports received during a campaign. | integer |
```

File: README.md [hunk 14]
```diff
@@ -21,7 +21,8 @@ campaign assessors to simplify the process of managing GoPhish campaigns.
 * `gophish-complete` - Completes a campaign in GoPhish and/or outputs a
   GoPhish campaign summary.
 * `gophish-export` - Exports all the data from an assessment within GoPhish
```
Comment (mcdonnnj) [inline README.md:23]: Please make an issue to replace all instances of `GoPhish` with `Gophish` in this project. We've slowly been making the change in other projects as we do updates and I don't want to scope creep this PR.

**Review by mcdonnnj — CHANGES_REQUESTED** [review]
I had some suggestions and a couple of issues I would like created. Thanks for your work!

**Review by jsf9k — APPROVED** [review]
:+1:

File: src/tools/gophish_export.py (hunk 10 above)
Comment (dav3r) [reply src/tools/gophish_export.py:289]: For a good explanation of why we no longer put f-strings in logging statements, check out [this post](https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/).

File: src/tools/gophish_export.py (hunk 12 above)
Comment (dav3r) [reply src/tools/gophish_export.py]: @mcdonnnj - I like the comment.  I'd only make a minor change to minimize any confusion:
```
# The "customer" field is a placeholder added for operator convenience when
# working with the JSON file created.
```

Comment (jsf9k) [reply src/tools/gophish_export.py]: This is a good point to clarify.  Thanks for pointing this out @mcdonnnj.

File: src/tools/gophish_export.py (hunk 10 above)
Comment (JCantu248) [reply src/tools/gophish_export.py:289]: See [here for issue](https://github.com/cisagov/gophish-tools/issues/34)

File: src/tools/gophish_export.py (hunk 11 above)
Comment (JCantu248) [reply src/tools/gophish_export.py]: Word changed in 3c54a97003c86809fdefc23a46a7ca562492e01c

File: src/tools/gophish_export.py (hunk 12 above)
Comment (JCantu248) [reply src/tools/gophish_export.py]: This is changed in 3c54a97003c86809fdefc23a46a7ca562492e01c

File: README.md (hunk 13 above)
Comment (JCantu248) [reply README.md]: Changes incorporated in cf4906f0b51ec10f488d8d013bed0d28003dd4ce

File: README.md (hunk 14 above)
Comment (JCantu248) [reply README.md:23]: See [here](https://github.com/cisagov/gophish-tools/issues/35) for the new issue

**Review by mcdonnnj — APPROVED** [review]
I believe you have addressed all of my concerns. Thanks for working with us to get this finished!
